crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -630,33 +630,48 @@ impl WritebackCache {
     async fn commit_transaction_outputs(
         &self,
         epoch: EpochId,
-        digest: TransactionDigest,
+        digests: &[TransactionDigest],
     ) -> SuiResult {
         fail_point_async!("writeback-cache-commit");
 
-        let DashMapEntry::Occupied(occupied) = self.dirty.pending_transaction_writes.entry(digest)
-        else {
-            panic!("Attempt to commit unknown transaction {:?}", digest);
-        };
-
-        let outputs = occupied.get().clone();
+        let mut all_outputs = Vec::with_capacity(digests.len());
+        for tx in digests {
+            let Some(outputs) = self
+                .dirty
+                .pending_transaction_writes
+                .get(tx)
+                .map(|o| o.clone())
+            else {
+                panic!("Attempt to commit unknown transaction {:?}", tx);
+            };
+            all_outputs.push(outputs);
+        }
 
         // Flush writes to disk before removing anything from dirty set. otherwise,
         // a cache eviction could cause a value to disappear briefly, even if we insert to the
         // cache before removing from the dirty set.
         self.store
-            .write_transaction_outputs(epoch, outputs.clone())
+            .write_transaction_outputs(epoch, &all_outputs)
             .await?;
 
-        // Cache transaction before removing entry from self.dirty to avoid
-        // unnecessary cache misses
-        self.cached
-            .transactions
-            .insert(digest, outputs.transaction.clone());
+        for (tx_digest, outputs) in digests.iter().zip(all_outputs.into_iter()) {
+            assert!(self
+                .dirty
+                .pending_transaction_writes
+                .remove(tx_digest)
+                .is_some());
+            self.remove_transaction_from_dirty_set(epoch, *tx_digest, &outputs);
+        }
 
-        // releases lock on pending_transaction_writes
-        occupied.remove();
+        Ok(())
+    }
 
+    fn remove_transaction_from_dirty_set(
```

Comment:
Might be better to name this something like "flush_transactions_from_dirty_to_cached" to indicate that it actually adds to the cache.